### ISSUE TITLE
docker module doesn't handle restarting containers correctly, get port conflicts

### ISSUE THREAD

**Issue by ryanwalls** [ansible/ansible-modules-core#1251]
Ansible version: v1.9.1-1 
Docker version 1.5.0, build a8a31ef/1.5.0

I want to just pull and image and restart it when it has changed.  To accomplish this I'm using `state: reloaded` flag.  

```
- name: start container
  docker:
    image: "{{ image }}"
    command: "grunt serve:{{env}}"
    registry: 'https://quay.io/v1/'
    username: 3dsim+ansible
    email: "{{email}}"
    password: "{{password}}"
    ports:
      - "{{hostPort}}:{{appPort}}"
    pull: always
    restart_policy: always
    state: reloaded
```

This works the first time I try to start the container.  Then later I'll run it and I'll get this error (this is full -vvvv output): 

```
TASK: [docker | start container] ********************************************** 
<52.24.100.78> ESTABLISH CONNECTION FOR USER: ec2-user
<52.24.119.34> ESTABLISH CONNECTION FOR USER: ec2-user
<52.24.100.78> REMOTE_MODULE docker password=VALUE_HIDDEN registry=https://quay.io/v1/ command='grunt serve:qa' state=reloaded email= restart_policy=always image=quay.io/3dsim/login pull=always username=3dsim+ansible
<52.24.119.34> REMOTE_MODULE docker password=VALUE_HIDDEN registry=https://quay.io/v1/ command='grunt serve:qa' state=reloaded email= restart_policy=always image=quay.io/3dsim/login pull=always username=3dsim+ansible
<52.24.119.34> EXEC ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/Users/ryan/.ansible/cp/ansible-ssh-%h-%p-%r" -o StrictHostKeyChecking=no -o IdentityFile="foo.pem" -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ec2-user -o ConnectTimeout=60 52.24.119.34 /bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1430786148.49-222303501072955 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1430786148.49-222303501072955 && echo $HOME/.ansible/tmp/ansible-tmp-1430786148.49-222303501072955'
<52.24.100.78> EXEC ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/Users/ryan/.ansible/cp/ansible-ssh-%h-%p-%r" -o StrictHostKeyChecking=no -o IdentityFile="foo.pem" -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ec2-user -o ConnectTimeout=60 52.24.100.78 /bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1430786148.49-93151108343614 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1430786148.49-93151108343614 && echo $HOME/.ansible/tmp/ansible-tmp-1430786148.49-93151108343614'
<52.24.100.78> PUT /var/folders/sr/b0lqysjn3hqd4jgqfxkmvxlw0000gn/T/tmpsHG_KN TO /home/ec2-user/.ansible/tmp/ansible-tmp-1430786148.49-93151108343614/docker
<52.24.119.34> PUT /var/folders/sr/b0lqysjn3hqd4jgqfxkmvxlw0000gn/T/tmpVIXrCa TO /home/ec2-user/.ansible/tmp/ansible-tmp-1430786148.49-222303501072955/docker
<52.24.100.78> EXEC ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/Users/ryan/.ansible/cp/ansible-ssh-%h-%p-%r" -o StrictHostKeyChecking=no -o IdentityFile="foo.pem" -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ec2-user -o ConnectTimeout=60 52.24.100.78 /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=hieixkgbkwngxilafwdurnxxmdwszasl] password: " -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-hieixkgbkwngxilafwdurnxxmdwszasl; LANG=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/bin/python /home/ec2-user/.ansible/tmp/ansible-tmp-1430786148.49-93151108343614/docker; rm -rf /home/ec2-user/.ansible/tmp/ansible-tmp-1430786148.49-93151108343614/ >/dev/null 2>&1'"'"''
<52.24.119.34> EXEC ssh -C -tt -vvv -o ControlMaster=auto -o ControlPersist=60s -o ControlPath="/Users/ryan/.ansible/cp/ansible-ssh-%h-%p-%r" -o StrictHostKeyChecking=no -o IdentityFile="foo.pem" -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ec2-user -o ConnectTimeout=60 52.24.119.34 /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=untwdijpdzlcvaejuynuanrpnfbxpgaj] password: " -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-untwdijpdzlcvaejuynuanrpnfbxpgaj; LANG=en_US.UTF-8 LC_CTYPE=en_US.UTF-8 /usr/bin/python /home/ec2-user/.ansible/tmp/ansible-tmp-1430786148.49-222303501072955/docker; rm -rf /home/ec2-user/.ansible/tmp/ansible-tmp-1430786148.49-222303501072955/ >/dev/null 2>&1'"'"''
failed: [52.24.100.78] => {"changed": true, "failed": true}
msg: Docker API Error: Cannot start container e08f61914d3891ea09cbf75e688722728da1c9ad2b09c0f832fc8935d4b85610: Bind for 0.0.0.0:9000 failed: port is already allocated
failed: [52.24.119.34] => {"changed": true, "failed": true}
msg: Docker API Error: Cannot start container 8a8e74b5dd5af59212edd61d3327b161af5aaa6b2c3f7eb3ebe5ba43b76bc4f3: Bind for 0.0.0.0:9000 failed: port is already allocated
```

Basically it seems that the currently running container is not stopped before it tries to start the new one.  (And `reload` shouldn't try to start a new one anyway because the image hasn't change.)

Not sure what combination of flags is causing this problem.  Perhaps the `restart_policy: always` is causing problems.  

If I ssh directly on the server, I can see the old container is still running: 

```
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS                    NAMES
acf14748861e        032b3a044545        "grunt serve:qa"    35 minutes ago      Up 35 minutes       0.0.0.0:9000->9000/tcp   reverent_wilson   
```

**Comment by ryanwalls** [comment]
UPDATE.   This still seems like a bug because the module says it can identify containers by command/image combination--which it doesn't seem to be doing in my case.  But I've worked around it for now by just specifying a `name:` parameter.

**Comment by dlip** [comment]
I'm having a similar inconsistency with the "state: reloaded" option. The documentation says:
"asserts that all matching containers are running and restarts any that have any images or configuration out of date"

But whenever I use this option every container gets restarted, even if the image and options haven't changed.

**Comment by gregdek** [comment]
cc: @cove @joshuaconner @softzilla @smashwilson

**Comment by VascoVisser** [comment]
> I'm having a similar inconsistency with the "state: reloaded" option. The documentation says:
> "asserts that all matching containers are running and restarts any that have any images or configuration out of date"
> But whenever I use this option every container gets restarted, even if the image and options haven't changed.

I can confirm this behavior with Ansible 1.9.1 and Docker version 1.7.0. 
verbose output gives: `"reload_reasons": "net (bridge => )"`

**Comment by VascoVisser** [comment]
verbose output gives:

**Comment by smashwilson** [comment]
Ah ha. The "net" parameter inconsistency was addressed by #1609, but it's likely that there are others. As a temporary workaround, you can add an explicit `net=bridge` parameter to your module invocation.

Really the long-term solution to issues like this is to have a comprehensive test suite that perturbs each module parameter and asserts that reloading happens or does not happen when expected, preferably run on CI somewhere, and (even better) against a build matrix of docker and docker-py versions. It's on my list, but I'm not sure how soon I'll have the time to set that up.

**Comment by wrouesnel** [comment]
Found another issue - if you manually set the log_driver, then the docker module will always do a reload trying to set it back to json-file, even if you actually didn't specify it at all (and journald remains enabled).

**Comment by ansibot** [comment]
@cove, @softzilla, @zfil, @ThomasSteinbach, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@cove, @softzilla, @zfil, @ThomasSteinbach, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by ansibot** [comment]
This issue was migrated to https://github.com/ansible/ansible/issues/30137
